File: 1029-vertical-order-traversal-of-a-binary-tree/vertical-order-traversal-of-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def verticalTraversal(self, root: Optional[TreeNode]) -> List[List[int]]:
        node_list=[]
        queue=deque([(root, 0,0)])
        while queue:
            node, row, col= queue.popleft()
            if node:
                node_list.append((col, row, node.val))
                queue.append((node.left, row+1, col-1))
                queue.append((node.right, row+1, col+1))
        node_list.sort()
        res=defaultdict(list) 
        for col, row, value in node_list:
            res[col].append(value)
        return [res[x] for x in sorted(res)]

        

File: 1029-vertical-order-traversal-of-a-binary-tree/vertical-order-traversal-of-a-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> verticalTraversal(TreeNode* root) {
        // Vector to hold (col, row, value)
        vector<tuple<int, int, int>> nodes;

        // BFS queue: node, row, col
        queue<tuple<TreeNode*, int, int>> q;
        q.push({root, 0, 0});

        while (!q.empty()) {
            auto [node, row, col] = q.front();
            q.pop();

            if (node) {
                nodes.push_back({col, row, node->val});
                if (node->left)
                    q.push({node->left, row + 1, col - 1});
                if (node->right)
                    q.push({node->right, row + 1, col + 1});
            }
        }

        // Sort nodes: first by col, then row, then value
        sort(nodes.begin(), nodes.end());

        vector<vector<int>> res;
        int prev_col = INT_MIN;

        for (auto [col, row, val] : nodes) {
            if (col != prev_col) {
                res.push_back({});
                prev_col = col;
            }
            res.back().push_back(val);
        }

        return res;
    }
};